be/src/exec/hash_join_components.cpp

```diff
@@ -155,4 +385,380 @@ ChunkPtr SingleHashJoinBuilder::convert_to_spill_schema(const ChunkPtr& chunk) c
+
+    size_t _fit_L2_cache_max_rows = 0;
+    size_t _fit_L3_cache_max_rows = 0;
+
+    size_t _L2_cache_size = 0;
+    size_t _L3_cache_size = 0;
+
+    size_t _pushed_chunks = 0;
+};
+
+AdaptivePartitionHashJoinBuilder::AdaptivePartitionHashJoinBuilder(HashJoiner& hash_joiner)
+        : HashJoinBuilder(hash_joiner) {
+    const size_t DEFAULT_L2_CACHE_SIZE = 1 * 1024 * 1024;
+    const size_t DEFAULT_L3_CACHE_SIZE = 32 * 1024 * 1024;
+    const auto& cache_sizes = CpuInfo::get_cache_sizes();
+    _L2_cache_size = cache_sizes[CpuInfo::L2_CACHE];
+    _L3_cache_size = cache_sizes[CpuInfo::L3_CACHE];
+    _L2_cache_size = _L2_cache_size ? _L2_cache_size : DEFAULT_L2_CACHE_SIZE;
+    _L3_cache_size = _L3_cache_size ? _L3_cache_size : DEFAULT_L3_CACHE_SIZE;
+}
+
+size_t AdaptivePartitionHashJoinBuilder::_estimated_row_size(const HashTableParam& param) const {
+    size_t estimated_each_row = 0;
+
+    for (auto* tuple : param.build_row_desc->tuple_descriptors()) {
+        for (auto slot : tuple->slots()) {
+            if (param.build_output_slots.contains(slot->id())) {
+                estimated_each_row += get_size_of_fixed_length_type(slot->type().type);
+                estimated_each_row += type_estimated_overhead_bytes(slot->type().type);
+            }
+        }
+    }
+
+    // for hash table bucket
+    estimated_each_row += 4;
+
+    return estimated_each_row;
+}
+
+// We could use a better estimation model.
+size_t AdaptivePartitionHashJoinBuilder::_estimated_probe_cost(const HashTableParam& param) const {
+    size_t size = 0;
+
+    for (auto* tuple : param.probe_row_desc->tuple_descriptors()) {
+        for (auto slot : tuple->slots()) {
+            if (param.probe_output_slots.contains(slot->id())) {
+                size += get_size_of_fixed_length_type(slot->type().type);
+                size += type_estimated_overhead_bytes(slot->type().type);
+            }
+        }
+    }
+    // we define probe cost is bytes size * 6
+    return size * 6;
+}
+
+size_t AdaptivePartitionHashJoinBuilder::_estimated_build_cost(size_t build_row_size, int cache_level) const {
+    if (cache_level == 0) {
+        return build_row_size / 2;
+    } else if (cache_level == 1) {
+        return build_row_size;
+    } else {
+        return build_row_size * 2;
+    }
+}
+
+void AdaptivePartitionHashJoinBuilder::_adjust_partition_rows(size_t build_row_size) {
+    _fit_L2_cache_max_rows = _L2_cache_size / build_row_size;
+    _fit_L3_cache_max_rows = _L3_cache_size / build_row_size;
+
+    // If the hash table is smaller than the L2 cache. we don't think partition hash join is needed.
+    _partition_join_min_rows = _fit_L2_cache_max_rows;
+    // If the hash table after partition can't be loaded to L3. we don't think partition hash join is needed.
+    _partition_join_max_rows = _fit_L3_cache_max_rows * _partition_num;
+
+    if (_probe_estimated_costs + _estimated_build_cost(build_row_size, 0) < _estimated_build_cost(build_row_size, 1)) {
```

Comment:
so _estimated_build_cost(build_row_size, 0) for loading in  L2 cost, _estimated_build_cost(build_row_size, 1) for loading in  L3 cost, _estimated_build_cost(build_row_size, 2) for loading in memory cost?

suggest to use symbol constant to differentiate these three cases, for examples:
1. auto constexpr int L2_CACHE_COST = 0;
2. auto constexpr int L3_CACHE_COST = 1;
3. auto constexpr int MEMORY_COST = 2;